source/s3_meta_request.c

```diff
@@ -2320,51 +2395,94 @@ struct aws_s3_meta_request_poll_write_result aws_s3_meta_request_poll_write(
         meta_request->synced_data.async_write.waker = waker;
         meta_request->synced_data.async_write.waker_user_data = user_data;
         result.is_pending = true;
-
     } else {
         /* write call is OK */
 
         /* If we don't already have a buffer, grab one from the pool. */
         if (meta_request->synced_data.async_write.buffered_data_ticket == NULL) {
-            /* NOTE: we acquire a forced-buffer because there's a risk of deadlock if we
-             * waited for a normal ticket reservation, respecting the pool's memory limit.
-             * (See "test_s3_many_async_uploads_without_data" for description of deadlock scenario) */
-            meta_request->synced_data.async_write.buffered_data = aws_s3_buffer_pool_acquire_forced_buffer(
-                meta_request->client->buffer_pool,
-                meta_request->part_size,
-                &meta_request->synced_data.async_write.buffered_data_ticket /*out_new_ticket*/);
-        }
 
-        /* Copy as much data as we can into the buffer */
-        struct aws_byte_cursor processed_data =
-            aws_byte_buf_write_to_capacity(&meta_request->synced_data.async_write.buffered_data, &data);
+            if (meta_request->synced_data.async_write.buffered_data_ticket == NULL &&
+                meta_request->synced_data.async_write.buffered_ticket_future == NULL) {
+                /* NOTE: we acquire a forced-buffer because there's a risk of deadlock if we
+                 * waited for a normal ticket reservation, respecting the pool's memory limit.
+                 * (See "test_s3_many_async_uploads_without_data" for description of deadlock scenario) */
 
-        /* Don't store EOF unless we've consumed all data */
-        if ((data.len == 0) && eof) {
-            meta_request->synced_data.async_write.eof = true;
-        }
+                struct aws_s3_buffer_pool_reserve_meta meta = {
+                    .size = meta_request->part_size,
+                    .can_block = true,
+                    .meta_request = meta_request,
+                    .client = meta_request->client};
 
-        /* This write makes us ready to send (EOF, or we have enough data now to send at least 1 part) */
-        if (meta_request->synced_data.async_write.eof ||
-            meta_request->synced_data.async_write.buffered_data.len == meta_request->part_size) {
+                meta_request->synced_data.async_write.buffered_ticket_future =
+                    aws_s3_buffer_pool_reserve(meta_request->client->buffer_pool, meta);
 
-            meta_request->synced_data.async_write.ready_to_send = true;
-            ready_to_send = true;
+                AWS_FATAL_ASSERT(meta_request->synced_data.async_write.buffered_ticket_future);
+            }
+
+            if (aws_future_s3_buffer_ticket_is_done(meta_request->synced_data.async_write.buffered_ticket_future)) {
```

Comment:
I am not following the logic here. If we are trying to asynchronously get a buffer and return pending while that call completes, we need to call the waker callback to let the user know that they should call write again.

As discussed offline, we should fix this or kill this API.

Reply:
updated. simplified logic back to if mem cannot be acquired request fails. eventually we can get rid of this altogether